go/src/koding/db/mongodb/modelhelper/stacktemplate.go

```diff
@@ -52,6 +54,39 @@ func GetStackTemplateFieldsByIds(ids []bson.ObjectId, fields []string) ([]*model
 	return stackTmpls, Mongo.Run(StackTemplateColl, query)
 }
 
+// ErrNoAccess is returned:
+//
+//   - by HasTemplateAccess function, when user has no access
+//     to the stack template
+//
+var ErrNoAccess = errors.New("no access for the requested resource")
+
+// HasTemplateAccess return non-nil error if the user is not allowed to access
+// the given stack template. It return nil error otherwise.
+func HasTemplateAccess(tmpl *models.StackTemplate, username string) error {
+	switch tmpl.AccessLevel {
+	case models.AccessPublic:
+		return nil // everyone has access
+	case models.AccessPrivate, "": // if AccessLevel is missing, we assume it's private
+		originID, err := GetAccountID(username)
+		if err != nil {
+			return ErrNoAccess
+		}
+
+		if originID != tmpl.OriginID {
+			return ErrNoAccess
+		}
+	case models.AccessGroup:
+		if ok, err := HasAnyRole(username, tmpl.Group, DefaultRoles...); err != nil || !ok {
+			return ErrNoAccess
+		}
+	default:
+		return errors.New("unrecognized access level: " + tmpl.AccessLevel)
+	}
+
+	return nil
```

Comment:
it's reachable for branches:
`models.AccessPrivate, ""`(err == nil,  and originID == tmpl.OriginID)
`models.AccessGroup`(err != nil || !ok evaluates to false)